upgrade/upgrade-scripts/rootfs-container

```diff
@@ -0,0 +1,263 @@
+#!/bin/bash
+#
+# Copyright 2019 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+. "${BASH_SOURCE%/*}/common.sh"
+
+CONTAINER=
+
+function get_dataset_snapshots() {
+	zfs list -rt snapshot -Hpo name "$1"
+}
+
+function get_snapshot_clones() {
+	zfs get clones -Hpo value "$1"
+}
+
+function delete() {
+	zfs list "rpool/ROOT/$CONTAINER/root" &>/dev/null ||
+		die "rootfs container '$CONTAINER' does not exist"
+
+	MOUNTED=$(zfs get mounted -Hpo value "rpool/ROOT/$CONTAINER/root")
+	[[ "$MOUNTED" == "no" ]] ||
+		die "cannot delete mounted rootfs container: '$CONTAINER'"
+
+	#
+	# The "data" and "home" datasets of a rootfs container may have
+	# been cloned as part of an upgrade. Thus, in order to delete
+	# this specific rootfs container, we need to promote any clones
+	# that exist. Otherwise we won't be able to destroy the snapshots
+	# for the "data" and "home" datasets, and thus can't destroy the
+	# datasets themselves.
+	#
+	for snap in \
```

Comment:
How would those clones get created?

While I tend to be defensive, and try to check for things like that.. I opted not to do that, because there isn't a way for those clones to be created. So, it's hard for me to justify why we'd check for clones, vs. anything else that we don't expect to exist and don't explicitly create.

But if you think it'd be more clear to check for clones, I can do that.

Reply:
If that container was an active `rootfs` at some point, then it could potentially have clones. While this will probably not happen in a normal use case, I feel like this check shouldn't cost anything but would safeguard against some potential issues, if we were to change the way some of the container creation works.

Reply:
> If that container was an active rootfs at some point, then it could potentially have clones.

i.e. if an "in-place" upgrade is concurrently running? the clone will only be created for in-place upgrade verification, right? I just want to make sure I'm understanding.

I guess it's also possible for an in-place upgrade container to have been manually created and currently running, which could allow for a clone exist, I think.

I'll add this check.

Reply:
I was able to get into this scenario by manually creating an upgrade container. Without a check, the failure message was bad (as might be expected):
```
$ sudo /var/dlpx-update/latest/rootfs-container delete delphix.JNHeZad
cannot destroy 'rpool/ROOT/delphix.JNHeZad': filesystem has dependent clones
use '-R' to destroy the following datasets:
rpool/ROOT/delphix.uE0noy5/root
rootfs-container: 'zfs destroy -r rpool/ROOT/delphix.JNHeZad' failed
```

With the check:
```
$ sudo /var/dlpx-update/latest/rootfs-container delete delphix.JNHeZad
rootfs-container: cannot delete rootfs container: 'delphix.JNHeZad', 'root' dataset clone exists: 'rpool/ROOT/delphix.uE0noy5/root'
```

Thanks for the feedback, this is much better.